[Feature](parquet-reader) Implements dict filter functionality parquet reader.

kaka11chen opened the pull request:
# Proposed changes

Implements dict filter functionality parquet reader to improve performance.


## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

kaka11chen committed:
[Feature](parquet-reader) Implements dict filter functionality in parquet reader.

kaka11chen commented:
run buildall

HappenLee commented on be/src/vec/data_types/data_type.cpp:
```diff
@@ -56,6 +56,7 @@ void IDataType::update_avg_value_size_hint(const IColumn& column, double& avg_va
 
 ColumnPtr IDataType::create_column_const(size_t size, const Field& field) const {
     auto column = create_column();
+    column->reserve(size);
```
const column nested column only need size == 1, do not need reserve `size`

kaka11chen replied:
Use `reserve(1)` instead.

HappenLee commented on be/src/vec/columns/column_dictionary.h:
```diff
@@ -166,21 +176,123 @@ class ColumnDictionary final : public COWHelper<IColumn, ColumnDictionary<T>> {
+                       << ", filter size: " << filter.size() << get_stack_trace();
+        }
+
+        const UInt8* filter_pos = filter.data();
```
filter return columnptr and size_t is similar， better to refactor

kaka11chen replied:
will rollback column_dictionary, so this case does not exists.

HappenLee commented on be/src/vec/exec/scan/vfile_scanner.cpp:
```diff
@@ -816,6 +876,11 @@ Status VFileScanner::_init_expr_ctxes() {
+
+    // TODO: It should can move to scan node to process.
+    if (_vconjunct_ctx && _vconjunct_ctx->root()) {
+        _split_conjuncts(_vconjunct_ctx->root());
```
which case `_vconjunct_ctx->root()` will be nullptr？

HappenLee replied:
which case `_vconjunct_ctx->root()` will be nullptr？

HappenLee commented on be/src/vec/exec/format/parquet/fix_length_dict_decoder.hpp:
```diff
@@ -434,13 +445,46 @@ class FixLengthDictDecoder<char*> final : public BaseDictDecoder {
+                          std::vector<int32_t>* dict_codes) override {
+        for (int i = 0; i < string_column->size(); ++i) {
+            StringRef dict_value = string_column->get_data_at(i);
+            dict_codes->emplace_back(_dict_value_to_code[dict_value]);
```
check `dict_codes` reverse before call `get_dict_codes`?

kaka11chen replied:
ok

HappenLee commented on be/src/vec/exec/format/parquet/vparquet_column_chunk_reader.h:
```diff
@@ -142,8 +143,23 @@ class ColumnChunkReader {
+                ->read_dict_values_to_column(doris_column);
+    }
+
+    Status get_dict_codes(const ColumnString* columnString, std::vector<int32_t>* dict_codes) {
```
`columnString` to `column_string`

kaka11chen replied:
ok

HappenLee commented on be/src/vec/exec/format/parquet/vparquet_column_reader.cpp:
```diff
@@ -384,6 +384,40 @@ Status ScalarColumnReader::_read_nested_column(ColumnPtr& doris_column, DataType
+
+Status ScalarColumnReader::get_dict_codes(const ColumnString* columnString,
+                                          std::vector<int32_t>* dict_codes) {
+    return _chunk_reader->get_dict_codes(columnString, dict_codes);
```
column_string

kaka11chen replied:
ok

HappenLee commented on be/src/vec/exec/format/parquet/vparquet_group_reader.cpp:
```diff
@@ -71,11 +97,136 @@ Status RowGroupReader::init(const FieldDescriptor& schema, std::vector<RowRange>
+
+    for (VExprContext* ctx : _slot_id_to_filter_conjuncts->at(slot_id)) {
+        const VExpr* root_expr = ctx->root();
+        if (root_expr->node_type() == TExprNodeType::FUNCTION_CALL) {
```
if the conjunt is `a > 10 is null ` ? the logic seems do the consider the case

HappenLee commented on be/src/vec/exec/format/parquet/vparquet_group_reader.cpp:
```diff
@@ -599,6 +668,236 @@ Status RowGroupReader::_filter_block_internal(Block* block,
+                    _column_readers[dict_filter_col_name]->convert_dict_column_to_string_column(
+                            dict_column);
+
+            size_t pos = block->get_position_by_name(dict_filter_col_name);
```
869 line search hash map by name， maybe better call `size_t pos = block->get_position_by_name(dict_filter_col_name);` in  869, only search hash map once time

kaka11chen committed:
1. Fix review issues.
2. use is_dict_filter flag & ColumnInt32 instead ColumnDictionary.
3. Rebase master.

kaka11chen commented:
run buildall

kaka11chen commented:
run buildall

HappenLee reviewed (approved):
LGTM
